lib/hotswap-queue.js

```diff
@@ -0,0 +1,60 @@
+const TICKS = 16
+
+module.exports = class HotswapQueue {
+  constructor () {
+    this.priorities = [[], [], []]
+  }
+
+  * pick (peer) {
+    for (let i = 0; i < this.priorities.length; i++) {
+      // try first one more than second one etc etc
+      let ticks = (this.priorities.length - i) * TICKS
+      const queue = this.priorities[i]
+
+      for (let j = 0; j < queue.length; j++) {
+        const r = j + Math.floor(Math.random() * queue.length - j)
+        const a = queue[j]
+        const b = queue[r]
+
+        if (r !== j) {
+          queue[(b.hotswap.index = j)] = b
+          queue[(a.hotswap.index = r)] = a
+        }
+
+        if (hasInflight(b, peer)) continue
+
+        yield b
+
+        if (--ticks <= 0) break
+      }
+    }
+  }
+
+  add (block) {
+    if (block.hotswap !== null) this.remove(block)
+    if (block.inflight.length === 0 || block.inflight.length >= 3) return
```

Comment:
I think `blocks.inflight.length` can never be 0: it's only called from `_sendBlockRequest`, after a req was pushed to the inflights, so I'm not sure what it protects against

Reply:
just a sanity check cause sep file

Reply:
cause it makes the queue selection explode, could be an assertion obvs